omnigan/trainer.py

```diff
@@ -1773,3 +1292,280 @@ def eval_images(self, mode, domain):
+                softmax_preds = softmax(pred, dim=1)
+                # Direct entropy minimization
+                loss = self.losses["G"]["tasks"]["s"]["minent"](softmax_preds)
+                loss *= self.opts.train.lambdas.G["s"]["minent"]
+                full_loss += loss
+
+                self.logger.losses.gen.task["s"]["minent"]["r"] = loss.item()
+
+        # Fool ADVENT discriminator
+        if self.opts.gen.s.use_advent:
+            if for_ == "D":
+                domain_label = domain
+                logger = {}
+                loss_func = self.losses["D"]["advent"]
+                pred = pred.detach()
+                weight = self.opts.train.lambdas.advent.adv_main
+            else:
+                domain_label = "s"
+                logger = self.logger.losses.gen.task["s"]["advent"]
+                loss_func = self.losses["G"]["tasks"]["s"]["advent"]
+                weight = self.opts.train.lambdas.G["s"]["advent"]
+
+            if for_ == "D" or domain == "r":
+                if softmax_preds is None:
+                    softmax_preds = softmax(pred, dim=1)
+                loss = loss_func(
+                    softmax_preds,
+                    self.domain_labels[domain_label],
+                    self.D["s"]["Advent"],
+                )
+                loss *= weight
+                full_loss += loss
+                logger[domain] = loss.item()
+
+                if for_ == "D":
+                    if self.opts.dis.s.gan_type == "GAN" or "WGAN_norm":
+                        pass
+                    elif self.opts.dis.s.gan_type == "WGAN":
+                        for p in self.D["s"]["Advent"].parameters():
+                            p.data.clamp_(
+                                self.opts.dis.s.wgan_clamp_lower,
+                                self.opts.dis.s.wgan_clamp_upper,
+                            )
+                    elif self.opts.dis.s.gan_type == "WGAN_gp":
+                        prob_need_grad = autograd.Variable(pred, requires_grad=True)
+                        d_out = self.D["s"]["Advent"](prob_need_grad)
+                        gp = get_WGAN_gradient(prob_need_grad, d_out)
+                        gp_loss = gp * self.opts.train.lambdas.advent.WGAN_gp
+                        full_loss += gp_loss
+                    else:
+                        raise NotImplementedError
+
+        return full_loss
+
+    def masker_m_loss(self, x, z, target, domain, for_="G"):
+        assert for_ in {"G", "D"}
+        assert domain in {"r", "s"}
+        self.assert_z_matches_x(x, z)
+        assert x.shape[0] == target.shape[0] if target is not None else True
+        full_loss = 0
+        # ? output features classifier
+        pred_logits = self.G.decoders["m"](z)
+        pred_prob = sigmoid(pred_logits)
+        pred_prob_complementary = 1 - pred_prob
+        prob = torch.cat([pred_prob, pred_prob_complementary], dim=1)
+        if domain == "s" and for_ == "G":
```

Comment:
Detail but lines from 1434 to 1471 could also be together in a big 'for_ == "G"'

Reply:
Done